src/libraries/Microsoft.Extensions.Http/src/DependencyInjection/HttpClientKeyedLifetime.cs

```diff
@@ -0,0 +1,89 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Net.Http;
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    internal class HttpClientKeyedLifetime
+    {
+        public static readonly HttpClientKeyedLifetime Disabled = new(null!, null!, null!);
+
+        public object ServiceKey { get; }
+        public ServiceDescriptor Client { get; }
+        public ServiceDescriptor Handler { get; }
+
+        public bool IsDisabled => ReferenceEquals(this, Disabled);
+
+        private HttpClientKeyedLifetime(object serviceKey, ServiceDescriptor client, ServiceDescriptor handler)
+        {
+            ServiceKey = serviceKey;
+            Client = client;
+            Handler = handler;
+        }
+
+        private HttpClientKeyedLifetime(object serviceKey, ServiceLifetime lifetime)
+        {
+            ThrowHelper.ThrowIfNull(serviceKey);
+            ServiceKey = serviceKey;
+            Client = ServiceDescriptor.DescribeKeyed(typeof(HttpClient), ServiceKey, CreateKeyedClient, lifetime);
+            Handler = ServiceDescriptor.DescribeKeyed(typeof(HttpMessageHandler), ServiceKey, CreateKeyedHandler, lifetime);
+        }
+
+        public HttpClientKeyedLifetime(ServiceLifetime lifetime) : this(KeyedService.AnyKey, lifetime) { }
+        public HttpClientKeyedLifetime(string name, ServiceLifetime lifetime) : this((object)name, lifetime) { }
+
+        public void AddRegistration(IServiceCollection services)
+        {
+            if (IsDisabled)
+            {
+                return;
+            }
+
+            services.Add(Client);
+            services.Add(Handler);
```

Comment:
Does this change the behavior of any old code that was solely relying on `IHttpClientFactory.CreateClient` or a typed client registration? If not, why are we concerned about breaking people by adding keyed registrations by default?

Reply:
Initially, there was a problem with third party service providers, see https://github.com/dotnet/runtime/pull/90272#issuecomment-1677722199

I've also just today discovered another problem https://github.com/dotnet/runtime/issues/102654#issuecomment-2234721322 -- as a workaround, I had to tamper with the empty-name "default" client registration, which is technically breaking https://github.com/dotnet/runtime/pull/104943/commits/e958285e19b53b60c4161df8983c9fb27571eed1

(Also, we initially thought we'd like to change the Typed client lifetime to align to the keyed client. So it was one of the reasons, even though at the moment I'm considering against that...)

Plus, I strongly believe we need to address scope mismatch problem and IDisposable tracking problem before proceeding with on-by-default, or we're just asking for trouble...